### ISSUE TITLE
Allow specifying FQDN in hostname resource

### ISSUE THREAD

**Issue by evandam** [chef/chef#10694]
### Describe the Enhancement:

The `hostname` resource seems to update `/etc/hosts` and drop the `hostname` property as the first record, so `hostname --fqdn` will always return that. It would be useful to have a separate `fqdn` property that would be set first.

Example:

```ruby
hostname 'aws-ubuntu-1804.kitchen' do
  aliases ['aws-ubuntu-1804.kitchen.example.com']
  ipaddress node['ipaddress']
end
```

Will result in an `/etc/hosts` like the following:

```
10.62.205.209 aws-ubuntu-1804.kitchen aws-ubuntu-1804.kitchen.example.com aws-ubuntu-1804
```

```
$ hostname -f
aws-ubuntu-1804.kitchen
```

### Describe the Need:

Something like an `fqdn` property that would set `/etc/hosts` correctly.

```ruby
hostname 'aws-ubuntu-1804.kitchen' do
  fqdn 'aws-ubuntu-1804.kitchen.example.com'
  ipaddress node['ipaddress']
end
```

```
10.62.205.209 aws-ubuntu-1804.kitchen.example.com aws-ubuntu-1804.kitchen aws-ubuntu-1804
```

```
$ hostname -f
aws-ubuntu-1804.kitchen.example.com
```

### Current Alternative

Alternative would be to skip the provided `hostname` resource and update `/etc/hosts` and run `hostnamectl set-hostname` by hand, unless there's a better option 😅

**Comment by phiggins** [comment]
Hi, I believe the `hostname` property is what you're looking for: https://docs.chef.io/resources/hostname/#properties

If I run your first example in an ubuntu vm I see the behavior you're describing. If afterwards I change it to look like this, I get what I believe is your desired behavior:

```
hostname 'aws-ubuntu-1804.kitchen' do
  hostname 'aws-ubuntu-1804.kitchen.example.com'
  ipaddress node['ipaddress']
end
```

Here's my session where I try this:

```
root@ubuntu-bionic:/vagrant/chef# hostname -f
ubuntu-bionic
root@ubuntu-bionic:/vagrant/chef# bundle exec chef-apply test_hostname.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * hostname[aws-ubuntu-1804.kitchen] action set
    * ohai[reload hostname] action nothing (skipped due to action :nothing)
    * execute[set hostname to aws-ubuntu-1804.kitchen] action run
      - execute /bin/hostname aws-ubuntu-1804.kitchen
    * file[/etc/hosts] action create
      - update content in file /etc/hosts from af8cff to f71241
      --- /etc/hosts    2020-04-02 23:52:33.124000000 +0000
      +++ /etc/.chef-hosts20201125-25622-cul9zk 2020-11-25 20:48:20.254000000 +0000
      @@ -8,5 +8,5 @@
       ff02::2  ip6-allrouters
       ff02::3  ip6-allhosts
       127.0.1.1        ubuntu-bionic   ubuntu-bionic
      -
      +10.0.2.15 aws-ubuntu-1804.kitchen aws-ubuntu-1804.kitchen.example.com aws-ubuntu-1804
    * execute[hostnamectl set-hostname aws-ubuntu-1804.kitchen] action run
      - execute hostnamectl set-hostname aws-ubuntu-1804.kitchen
    * ohai[reload hostname] action reload/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: already initialized constant MAC_MATCH
/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: previous definition of MAC_MATCH was here

      - re-run ohai and merge results into node attributes
    * ohai[reload hostname] action reload/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: already initialized constant MAC_MATCH
/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: previous definition of MAC_MATCH was here

      - re-run ohai and merge results into node attributes

  * hostname[aws-ubuntu-1804.kitchen] action nothing (skipped due to action :nothing)
root@ubuntu-bionic:/vagrant/chef#
root@ubuntu-bionic:/vagrant/chef#
root@ubuntu-bionic:/vagrant/chef#
root@ubuntu-bionic:/vagrant/chef# hostname -f
aws-ubuntu-1804.kitchen
root@ubuntu-bionic:/vagrant/chef# vim test_hostname.rb
root@ubuntu-bionic:/vagrant/chef# bundle exec chef-apply test_hostname.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * hostname[aws-ubuntu-1804.kitchen] action set
    * ohai[reload hostname] action nothing (skipped due to action :nothing)
    * execute[set hostname to aws-ubuntu-1804.kitchen.example.com] action run
      - execute /bin/hostname aws-ubuntu-1804.kitchen.example.com
    * file[/etc/hosts] action create
      - update content in file /etc/hosts from f71241 to bfcb4b
      --- /etc/hosts    2020-11-25 20:48:20.254000000 +0000
      +++ /etc/.chef-hosts20201125-26012-wvrqpn 2020-11-25 20:50:41.642000000 +0000
      @@ -8,5 +8,5 @@
       ff02::2  ip6-allrouters
       ff02::3  ip6-allhosts
       127.0.1.1        ubuntu-bionic   ubuntu-bionic
      -10.0.2.15 aws-ubuntu-1804.kitchen aws-ubuntu-1804.kitchen.example.com aws-ubuntu-1804
      +10.0.2.15 aws-ubuntu-1804.kitchen.example.com aws-ubuntu-1804
    * execute[hostnamectl set-hostname aws-ubuntu-1804.kitchen.example.com] action run
      - execute hostnamectl set-hostname aws-ubuntu-1804.kitchen.example.com
    * ohai[reload hostname] action reload/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: already initialized constant MAC_MATCH
/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: previous definition of MAC_MATCH was here

      - re-run ohai and merge results into node attributes
    * ohai[reload hostname] action reload/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: already initialized constant MAC_MATCH
/root/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/ohai-8d1ae4294224/lib/ohai/plugins/eucalyptus.rb:32: warning: previous definition of MAC_MATCH was here

      - re-run ohai and merge results into node attributes

  * hostname[aws-ubuntu-1804.kitchen] action nothing (skipped due to action :nothing)
root@ubuntu-bionic:/vagrant/chef# hostname -f
aws-ubuntu-1804.kitchen.example.com
```

**Comment by evandam** [comment]
Thanks @phiggins, but I think it's a little different. Setting `hostname` will update the FQDN, but it will still match the hostname.

The goal is to have something that results in the following:

```
# hostname
aws-ubuntu-1804.kitchen

# hostname -f
aws-ubuntu-1804.kitchen.example.com
```

**Comment by tas50** [comment]
@evan2645 This seems like a legit need. It's a bit different than just setting the hostname.

**Comment by tas50** [comment]
This has been merged to master and will ship in 17.0
